### PR TITLE
implement User,secure logic

### PR DESCRIPTION
**Pull request by Roma1990Melnichuk**

### CODE REVIEW

File: src/main/java/com/bookstore/controller/AuthController.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/auth")
+public class AuthController {
+
+    private final UserService userService;
+
+    @PostMapping("/registration")
+    public UserResponseDto registerUser(@Valid @RequestBody UserRegistrationRequestDto requestDto)
```
Comment (JJJazl) [inline src/main/java/com/bookstore/controller/AuthController.java]: Please, add the appropriate status code for POST method

File: src/main/java/com/bookstore/exception/RegistrationException.java [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package com.bookstore.exception;
+
+public class RegistrationException extends Exception {
```
Comment (JJJazl) [inline src/main/java/com/bookstore/exception/RegistrationException.java]:
```suggestion
public class RegistrationException extends RuntimeException {
```

File: src/main/java/com/bookstore/service/UserService.java [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package com.bookstore.service;
+
+import com.bookstore.dto.UserRegistrationRequestDto;
+import com.bookstore.dto.UserResponseDto;
+import com.bookstore.exception.RegistrationException;
+
+public interface UserService {
+    UserResponseDto register(UserRegistrationRequestDto requestDto)
+            throws RegistrationException;
```
Comment (JJJazl) [inline src/main/java/com/bookstore/service/UserService.java]:
```suggestion
    UserResponseDto register(UserRegistrationRequestDto requestDto);
```

File: src/main/java/com/bookstore/controller/AuthController.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+            throws RegistrationException {
```
Comment (JJJazl) [inline src/main/java/com/bookstore/controller/AuthController.java]:
```suggestion
    public UserResponseDto registerUser(@Valid @RequestBody UserRegistrationRequestDto requestDto) {
```

File: src/main/java/com/bookstore/service/UserServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+@RequiredArgsConstructor
+@Service
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+    private final PasswordEncoder passwordEncoder;
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto requestDto)
```
Comment (JJJazl) [inline src/main/java/com/bookstore/service/UserServiceImpl.java]:
```suggestion
    @Override
    @Transactional
    public UserResponseDto register(UserRegistrationRequestDto requestDto)
```

File: src/main/resources/db/changelog/changes/02-create-users-table.yaml [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        - createTable:
+            tableName: users
+            columns:
+              - column:
+                  name: id
+                  type: bigint
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+                    nullable: false
```
Comment (JJJazl) [inline src/main/resources/db/changelog/changes/02-create-users-table.yaml]: Let's have an explicit constraint name for PK (primary key)

**Review by JJJazl — CHANGES_REQUESTED** [review]

File: src/main/java/com/bookstore/service/UserServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+@Service
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+    private final PasswordEncoder passwordEncoder;
+
+    @Override
+    @Transactional
+    public UserResponseDto register(UserRegistrationRequestDto requestDto)
+            throws RegistrationException {
```
Comment (okuzan) [inline src/main/java/com/bookstore/service/UserServiceImpl.java]: not fixed, it should be unchecked exception

File: src/main/java/com/bookstore/controller/AuthController.java [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/auth")
+public class AuthController {
+
```
Comment (okuzan) [inline src/main/java/com/bookstore/controller/AuthController.java]:
```suggestion
```

**Review by okuzan — CHANGES_REQUESTED** [review]

File: pom.xml [hunk 9]
```diff
@@ -23,6 +23,10 @@
 		<checkstyle.plugin.version>3.3.0</checkstyle.plugin.version>
 		<lombok.mapstruct.binding.version>0.1.0 </lombok.mapstruct.binding.version>
 		<springdoc-openapi-starter-webmvc-ui.version>2.6.0 </springdoc-openapi-starter-webmvc-ui.version>
+		<commons-beanutils.version>1.9.4 </commons-beanutils.version>
```
Comment (Elena-Bruyako) [inline pom.xml]: remove this dependency

File: pom.xml [hunk 10]
```diff
@@ -78,6 +82,27 @@
...
 		</dependency>
+		<dependency>
+			<groupId>commons-beanutils</groupId>
+			<artifactId>commons-beanutils</artifactId>
+			<version>${commons-beanutils.version}</version>
+		</dependency>
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-security</artifactId>
+			<version>${spring-boot-starter-security.version}</version>
```
Comment (Elena-Bruyako) [inline pom.xml]: read about starters

```suggestion
```

File: pom.xml [hunk 11]
```diff
@@ -78,6 +82,27 @@
...
+		</dependency>
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-web</artifactId>
+			<version>${spring-boot-starter-web.version}</version>
```
Comment (Elena-Bruyako) [inline pom.xml]: don't need to specify the version, spring-boot-parent already provides it

```suggestion
```

File: pom.xml [hunk 12]
```diff
@@ -78,6 +82,27 @@
...
+		</dependency>
+		<dependency>
+			<groupId>org.springframework.security</groupId>
+			<artifactId>spring-security-core</artifactId>
```
Comment (Elena-Bruyako) [inline pom.xml]: remove dependency

File: src/main/java/com/bookstore/controller/AuthController.java [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/auth")
+public class AuthController {
+    private final UserService userService;
+
+    @PostMapping("/registration")
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/controller/AuthController.java:22]: add Swagger annotations to controller

File: src/main/java/com/bookstore/entity/User.java [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+    @Column(nullable = false)
+    private String password;
+
+    @Column(nullable = false, name = "first_name")
+    private String firstName;
+
+    @Column(nullable = false, name = "last_name")
+    private String lastName;
+
+    @Column(name = "shipping_address")
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/entity/User.java]: Don't need to add `name` attribute to Column annotation. Hibernate automatically creates column name in snake_case format

```suggestion
    @Column(nullable = false)
    private String firstName;
    @Column(nullable = false)
    private String lastName;
```

File: src/main/java/com/bookstore/service/UserServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+@Service
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+    private final PasswordEncoder passwordEncoder;
+
+    @Override
+    @Transactional
+    public UserResponseDto register(UserRegistrationRequestDto requestDto) {
+        if (userRepository.findByEmail(requestDto.getEmail()).isPresent()) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/service/UserServiceImpl.java]:
```suggestion
        if (userRepository.existsByEmail(requestDto.getEmail())) {
```

File: src/main/java/com/bookstore/service/UserServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+            throw new RegistrationException("User with this email already exists");
+        }
+        User user = userMapper.toEntity(requestDto);
+        user.setPassword(passwordEncoder.encode(requestDto.getPassword()));
+        User savedUser = userRepository.save(user);
+        return userMapper.toUserResponseDto(savedUser);
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/service/UserServiceImpl.java]:
```suggestion
        userRepository.save(user);
        return userMapper.toUserResponseDto(user);
```

File: src/main/java/com/bookstore/dto/UserRegistrationRequestDto.java [hunk 17]
```diff
@@ -0,0 +1,24 @@
+package com.bookstore.dto;
+
+import com.bookstore.validation.FieldMatch;
+import jakarta.validation.constraints.Email;
+import jakarta.validation.constraints.NotBlank;
+import lombok.Data;
+
+@Data
+@FieldMatch(first = "password", second = "repeatPassword",
+        patternOf = "^(?=.*[A-Z])(?=.*[a-z])(?=.*\\d)[A-Za-z\\d]{8,}$")
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/dto/UserRegistrationRequestDto.java]: replace this pattern as a constant to validator

File: src/main/java/com/bookstore/validation/FieldMatchValidator.java [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+            return true;
+        }
+        try {
+            Field firstField = value.getClass().getDeclaredField(firstFieldName);
+            Field secondField = value.getClass().getDeclaredField(secondFieldName);
+            firstField.setAccessible(true);
+            secondField.setAccessible(true);
+            String firstFieldValue = (String) firstField.get(value);
+            String secondFieldValue = (String) secondField.get(value);
+            return (firstFieldValue == null && secondFieldValue == null)
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/validation/FieldMatchValidator.java]: you can use Objects.equals()

File: src/main/resources/application.properties [hunk 19]
```diff
@@ -1,6 +1,7 @@
 spring.datasource.url=jdbc:mysql://localhost:3306/book_store
 spring.datasource.username=root
 spring.datasource.password=test
-spring.jpa.hibernate.ddl-auto= validate
+spring.jpa.hibernate.ddl-auto= update
```
Comment (Elena-Bruyako) [inline src/main/resources/application.properties]: revert
```suggestion
spring.jpa.hibernate.ddl-auto=validate
```

File: src/main/resources/db/changelog/changes/02-create-users-table.yaml [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+              - column:
+                  name: last_name
+                  type: varchar(255)
+                  constraints:
+                    nullable: false
+              - column:
+                  name: shipping_address
+                  type: varchar(255)
+              - column:
+                  name: is_deleted
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/02-create-users-table.yaml]: didn't find isDeleted column in User entity

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/com/bookstore/controller/AuthController.java [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+    @Operation(
+            summary = "Register a new user",
+            description = "Registers a new user with the "
+                    + "provided information and returns a confirmation of successful registration."
+    )
+    @ResponseStatus(HttpStatus.CREATED)
+
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/controller/AuthController.java]: remove empty line

File: src/main/java/com/bookstore/controller/AuthController.java [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+    public UserResponseDto registerUser(@Valid @RequestBody UserRegistrationRequestDto requestDto) {
+
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/controller/AuthController.java]: same here

File: src/main/java/com/bookstore/entity/User.java [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+    @Column(nullable = false)
+    private String password;
+
+    @Column(nullable = false)
+    private String firstName;
+
+    @Column(nullable = false)
+    private String lastName;
+
+    @Column(name = "shipping_address")
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/entity/User.java]:
```suggestion
```

File: src/main/java/com/bookstore/validation/FieldMatchValidator.java [hunk 24]
```diff
@@ -0,0 +1,54 @@
...
+import jakarta.validation.ConstraintValidator;
+import jakarta.validation.ConstraintValidatorContext;
+import java.lang.reflect.Field;
+import java.util.Objects;
+import java.util.regex.Pattern;
+
+public class FieldMatchValidator implements ConstraintValidator<FieldMatch, Object> {
+    private String firstFieldName;
+    private String secondFieldName;
+    private String pattern;
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/validation/FieldMatchValidator.java]:
```suggestion
    public static final String PASSWORD_PATTERN =
            "^(?=.*[A-Z])(?=.*[a-z])(?=.*\\d)[A-Za-z\\d]{8,}$";
```

File: src/main/java/com/bookstore/validation/PasswordValidator.java [hunk 25]
```diff
@@ -0,0 +1,6 @@
+package com.bookstore.validation;
+
+public class PasswordValidator {
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/validation/PasswordValidator.java]: remove

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by ahoienko — APPROVED** [review]
Good job!

**Pull request merged**